### ISSUE TITLE
An issue with unclear clothing quality during training

### ISSUE THREAD

**Issue by ZhangLei999** [Zheng-Chong/CatVTON#102]
I trained with 15W+ datasets, and some of the clothes I trained had blurry details.Taking the following picture as an example, the edges of the clothes appear blurry, and the texture details of the clothes are not very accurate. 

![Image](https://github.com/user-attachments/assets/11a9824b-905a-4f0d-ab1b-0899a7d4c339)

**Comment by ZhangLei999** [comment]
Using the same data, I trained with OOT and the texture was very clear.

![Image](https://github.com/user-attachments/assets/68b961eb-2beb-4490-8214-c1a0b6267431)

What is the reason? I have used all the training strategies mentioned in the paper. The metrics I trained on the VITONHD dataset are very similar to the metrics in the paper.

**Comment by ruanruano** [comment]
> Using the same data, I trained with OOT and the texture was very clear.
> 
> ![Image](https://github.com/user-attachments/assets/68b961eb-2beb-4490-8214-c1a0b6267431)
> 
> What is the reason? I have used all the training strategies mentioned in the paper. The metrics I trained on the VITONHD dataset are very similar to the metrics in the paper.

Hello, I also want to train and reproduce the code results of ootd and catvton recently. Can you share your training code? I will be very honored if you reply to me. This is my email address 1742520910@qq.com

**Comment by jasnei** [comment]
> I trained with 15W+ datasets, and some of the clothes I trained had blurry details.Taking the following picture as an example, the edges of the clothes appear blurry, and the texture details of the clothes are not very accurate.
> 
> ![Image](https://github.com/user-attachments/assets/11a9824b-905a-4f0d-ab1b-0899a7d4c339)

Hi, I had tried many ways, but I could reproduce the training result. Could you please kindly share your training script. Here is my email: albertyuang@gmail.com

Many Thanks

**Comment by Heimdall-Nss** [comment]
> Using the same data, I trained with OOT and the texture was very clear.
> 
> ![Image](https://github.com/user-attachments/assets/68b961eb-2beb-4490-8214-c1a0b6267431)
> 
> What is the reason? I have used all the training strategies mentioned in the paper. The metrics I trained on the VITONHD dataset are very similar to the metrics in the paper.

Hello, I am also trying to reproduce the method, but I cannot achieve the results in the paper. Can you please share your training script? This is my email: dundun.rebuild@gmail.com

Thank you very much

**Comment by syedmomin** [comment]
@jasnei @ZhangLei999 @Heimdall-Nss @ruanruano 
Hello, I am also trying to train model in kaggle but I cannot achieve the results. Can you please guide me how did you get here? This is my email: [syedmomin168@gmail.com](mailto:dundun.rebuild@gmail.com)

Thank you very much

**Comment by wwwjh333** [comment]
> Using the same data, I trained with OOT and the texture was very clear.
> 
> ![Image](https://github.com/user-attachments/assets/68b961eb-2beb-4490-8214-c1a0b6267431)
> 
> What is the reason? I have used all the training strategies mentioned in the paper. The metrics I trained on the VITONHD dataset are very similar to the metrics in the paper.

Hello,我最近复现了catvton的训练代码，我用了文中提到的所有策略，但是仍然会有生成图像质量纹理很差的效果，也许可以参考您的训练代码参考吗，这是我的邮箱: wujunhao333@gmail.com

**Comment by all2xii** [comment]
@ZhangLei999 @jasnei @syedmomin @Heimdall-Nss @wwwjh333  Your replicated training code is an impressive piece of work! I’m deeply interested in the implementation details of your training code. Would you be willing to share the complete code for educational or collaborative purposes? I’d be happy to discuss compensating you for your work or sponsoring your efforts.this is my email:lzx1908249431@gmail.com

**Comment by Thuw-ortho** [comment]
> I trained with 15W+ datasets, and some of the clothes I trained had blurry details.Taking the following picture as an example, the edges of the clothes appear blurry, and the texture details of the clothes are not very accurate.
> 
> ![Image](https://github.com/user-attachments/assets/11a9824b-905a-4f0d-ab1b-0899a7d4c339)

Hello, 

Thank you for sharing your results. I've also trying to reproduce the method but haven't been able to achieve results as good as CatVTON. If you're open to it, could you kindly share your training code? I'd really appreciate it. 
You can reach me at: minhthu.le@ortho.fashion

Thank you very much!

**Comment by Harsh-Kesharwani** [comment]
> I trained with 15W+ datasets, and some of the clothes I trained had blurry details.Taking the following picture as an example, the edges of the clothes appear blurry, and the texture details of the clothes are not very accurate.
> 
> ![Image](https://github.com/user-attachments/assets/11a9824b-905a-4f0d-ab1b-0899a7d4c339)
Thank you for sharing your results.
can you please share training script. As i have tried alot but it is not giving better results. can you please connect with me: harshkesherwani3@gmail.com

**Comment by Harsh-Kesharwani** [comment]
@ZhangLei999 can you please check whether below loss computation is correct or not. I follow dream strategy as mention in research paper. but is input to diffusion model is correct ? here diffusion model is my unet which i have build from scratch and load pre-trained weights of sd-v1.5 inpainting.
 
```
def compute_loss(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
        """Compute MSE loss for denoising with DREAM strategy"""
        batch_size = batch['person_image'].shape[0]
        
        # Prepare images
        person_images = batch['person_image'].to(self.device)
        garment_images = batch['garment_image'].to(self.device) 
        masks = batch['mask'].to(self.device)
        
        # Apply masks to person images (mask-based training)
        masked_person = person_images * (masks < 0.5)
        
        # Encode images to latent space
        encoder = self.models['encoder']
        
        with torch.cuda.amp.autocast(enabled=self.use_mixed_precision):
            # Get latents - concatenate person and garment spatially as per paper
            person_latent = compute_vae_encodings(person_images, encoder, self.device)
            masked_person_latent = compute_vae_encodings(masked_person, encoder, self.device)
            garment_latent = compute_vae_encodings(garment_images, encoder, self.device)
            
            # Apply CFG dropout to garment latent
            garment_latent = self._apply_cfg_dropout(garment_latent)
            
            # Concatenate spatially as per paper: Xi = ε(Ii©Ig) where © is spatial concatenation
            # Input person (masked) and garment concatenated along width
            input_latent = torch.cat([masked_person_latent, garment_latent], dim=-1)  # Concatenate along width
            target_latent = torch.cat([person_latent, garment_latent], dim=-1)  # Target includes unmasked person
            
            # Prepare mask condition - interpolate mask to latent size
            mask_latent = F.interpolate(masks.unsqueeze(1) if masks.ndim == 3 else masks, size=person_latent.shape[-2:], mode="nearest")
            # Create all-zero mask for garment area
            zero_mask = torch.zeros_like(mask_latent)
            mask_condition = torch.cat([mask_latent, zero_mask], dim=-1)  # Concatenate along width
            
            # Sample random timesteps
            timesteps = torch.randint(
                0, self.scheduler.num_train_timesteps, (batch_size,), 
                device=self.device, dtype=torch.long
            )
            
            # Sample noise
            noise = torch.randn_like(target_latent)
            
            # Add noise to target latents (forward diffusion)
            noisy_latents = self.scheduler.add_noise(target_latent, timesteps, noise)
            
            # DREAM strategy implementation
            if self.dream_lambda > 0:
                # First forward pass to get initial prediction
                model_input = torch.cat([noisy_latents, mask_condition], dim=1)
                time_embeddings = []
                for t in timesteps:
                    time_emb = get_time_embedding(t.item()).to(self.device)
                    time_embeddings.append(time_emb)
                time_embeddings = torch.stack(time_embeddings).squeeze(1)
                
                # Get initial noise prediction
                diffusion_model = self.models['diffusion']
                with torch.no_grad():
                    epsilon_theta = diffusion_model(model_input, time_embeddings)
                
                # Apply DREAM: zˆt = √αt*z0 + √(1-αt)*(ε + λ*εθ)
                alphas_cumprod = self.scheduler.alphas_cumprod.to(device=target_latent.device, dtype=target_latent.dtype)
                sqrt_alpha_prod = alphas_cumprod[timesteps] ** 0.5
                sqrt_one_minus_alpha_prod = (1 - alphas_cumprod[timesteps]) ** 0.5
                
                # Reshape for broadcasting
                sqrt_alpha_prod = sqrt_alpha_prod.view(-1, 1, 1, 1)
                sqrt_one_minus_alpha_prod = sqrt_one_minus_alpha_prod.view(-1, 1, 1, 1)
                
                # DREAM noise combination
                dream_noise = noise + self.dream_lambda * epsilon_theta
                dream_noisy_latents = sqrt_alpha_prod * target_latent + sqrt_one_minus_alpha_prod * dream_noise
                
                # Use DREAM noisy latents for training
                model_input = torch.cat([dream_noisy_latents, mask_condition], dim=1)
                predicted_noise = diffusion_model(model_input, time_embeddings)
                
                # DREAM loss: |(ε + λεθ) - εθ(ẑt, t)|²
                loss = F.mse_loss(predicted_noise, dream_noise)
            else:
                # Standard training without DREAM
                model_input = torch.cat([noisy_latents, mask_condition], dim=1)
                time_embeddings = []
                for t in timesteps:
                    time_emb = get_time_embedding(t.item()).to(self.device)
                    time_embeddings.append(time_emb)
                time_embeddings = torch.stack(time_embeddings).squeeze(1)
                
                diffusion_model = self.models['diffusion']
                predicted_noise = diffusion_model(model_input, time_embeddings)
                
                # Standard MSE loss
                loss = F.mse_loss(predicted_noise, noise)
        
        return loss
```

**Comment by baihuple** [comment]
@ZhangLei999 @jasnei @syedmomin @Heimdall-Nss @wwwjh333 Your replicated training code is an impressive piece of work! I’m deeply interested in the implementation details of your training code. Would you be willing to share the complete code for educational or collaborative purposes? I’d be happy to discuss compensating you for your work or sponsoring your efforts.this is my email:3259482249@qq.com

**Comment by rlawldud53** [comment]
@ZhangLei999 @jasnei @syedmomin @Heimdall-Nss @wwwjh333 Your replicated training code is an impressive piece of work! I’m deeply interested in the implementation details of your training code. Would you be willing to share the complete code for educational or collaborative purposes? I’d be happy to discuss compensating you for your work or sponsoring your efforts.this is my email: kplove01@kaist.ac.kr 

Thank you for your work !! 🙏🙏🙏🙏
